File: app/src/main/java/me/pwcong/jpstart/mvp/model/JPStartFragmentModelImpl.java
package me.pwcong.jpstart.mvp.model;

import java.util.List;

import me.pwcong.jpstart.conf.Constants;
import me.pwcong.jpstart.manager.DBManager;
import me.pwcong.jpstart.mvp.bean.JPItem;
import rx.Observable;
import rx.Subscriber;
import rx.android.schedulers.AndroidSchedulers;
import rx.schedulers.Schedulers;

/**
 * Created by Pwcong on 2016/9/27.
 */

public class JPStartFragmentModelImpl implements BaseModel.JPStartFragmentModel {

    @Override
    public void getData(final int category, Subscriber<List<JPItem>> subscriber) {
        Observable.create(new Observable.OnSubscribe<List<JPItem>>() {

            @Override
            public void call(Subscriber<? super List<JPItem>> subscriber) {

                List<JPItem> list;

                subscriber.onStart();

                switch (category) {

                    case Constants.CATEGORY_QINGYIN:
                        list = DBManager.getInstance().getQingYin();
                        break;
                    case Constants.CATEGORY_ZHUOYIN:
                        list = DBManager.getInstance().getZhuoYin();
                        break;
                    case Constants.CATEGORY_AOYIN:
                        list = DBManager.getInstance().getAoYin();
                        break;
                    default:
                        list = null;
                        break;

                }

                if (list == null) {
                    subscriber.onError(new Exception());
                } else {
                    subscriber.onNext(list);
                }

                subscriber.onCompleted();
            }
        }).observeOn(AndroidSchedulers.mainThread()).subscribeOn(Schedulers.newThread()).subscribe(subscriber);
    }
}

File: app/src/main/java/com/github/pwcong/jpstart/mvp/model/impl/JPStartFragmentModelImpl.kt
package com.github.pwcong.jpstart.mvp.model.impl

import com.github.pwcong.jpstart.constants.Constants
import com.github.pwcong.jpstart.manager.DBManager
import com.github.pwcong.jpstart.mvp.bean.JPItem
import com.github.pwcong.jpstart.mvp.model.BaseModel.JPStartFragmentModel
import rx.Observable
import rx.Subscriber
import rx.android.schedulers.AndroidSchedulers
import rx.schedulers.Schedulers

class JPStartFragmentModelImpl : JPStartFragmentModel {
    override fun getData(category: Int, subscriber: Subscriber<List<JPItem>>) {
        Observable.create { subscriber ->
            subscriber.onStart()
            val list: List<JPItem>? = when (category) {
                Constants.CATEGORY_QINGYIN -> DBManager.getInstance().getQingYin()

                Constants.CATEGORY_ZHUOYIN -> DBManager.getInstance().getZhuoYin()

                Constants.CATEGORY_AOYIN -> DBManager.getInstance().getAoYin()

                else -> null
            }
            if (list == null) {
                subscriber.onError(Exception())
            } else {
                subscriber.onNext(list)
            }
            subscriber.onCompleted()
        }.observeOn(AndroidSchedulers.mainThread()).subscribeOn(Schedulers.newThread())
            .subscribe(subscriber)
    }
}